File: services/terminal-relay/src/framing.ts
/**
 * Wire framing of the agent leg: `[tag:1][length:4 big-endian][payload]`.
 *
 * The same bytes are produced by the Python PTY agent, so the encoders here are the
 * TypeScript half of a cross-language contract: no field order, padding or encoding may
 * change without changing the agent and the interop golden vectors together.
 */

export const FRAME_TAGS = {
  AGENT_HELLO: 0x01,
  HELLO_ACK: 0x02,
  OPEN: 0x10,
  OPEN_OK: 0x11,
  OPEN_ERR: 0x12,
  STDIN: 0x20,
  STDOUT: 0x21,
  RESIZE: 0x22,
  /** DA/DSR produced by the emulator; never human keyboard or paste. */
  TERMINAL_RESPONSE: 0x23,
  /** Per-session output credit: does not pause the agent's multiplexed socket. */
  PAUSE_OUTPUT: 0x24,
  RESUME_OUTPUT: 0x25,
  CLOSE: 0x30,
  CLOSED: 0x31,
  PING: 0x40,
  PONG: 0x41,
  // Governance file reads. A standalone transaction, not a session, so it does not reuse OPEN.
  // Listing them here is what stops the agent's reply from being an `unknown frame tag`: the
  // decoder treats an unknown tag as a violation and drops the ENTIRE connection, with every
  // open terminal on top.
  READ: 0x50,
  READ_OK: 0x51,
  READ_ERR: 0x52,
  READ_DATA: 0x53,
  /** Governed write v1. Sent only if the hello advertises `write_governance_v1`. */
  WRITE: 0x54,
  WRITE_DATA: 0x55,
  WRITE_OK: 0x56,
  WRITE_ERR: 0x57,
  /** Releases an incomplete write when its HTTP request expires or closes. */
  WRITE_CANCEL: 0x58,
  /** Multi-file profile: one transaction, full preflight, and rollback on the agent. */
  WRITE_BATCH: 0x59,
  WRITE_BATCH_DATA: 0x5a,
  WRITE_BATCH_OK: 0x5b,
  WRITE_BATCH_ERR: 0x5c,
  WRITE_BATCH_CANCEL: 0x5d,
  /** Unambiguous close of a successful read; READ_ERR is already terminal on its own. */
  READ_DONE: 0x5e
} as const;

export type FrameTag = (typeof FRAME_TAGS)[keyof typeof FRAME_TAGS];

/** A frame payload never exceeds this; a longer declared length is a protocol violation. */
export const MAX_FRAME_PAYLOAD_BYTES = 65_536;
/**
 * DATA payloads carry the session UUID (with dashes) as 36 leading ASCII bytes. READ_DATA and
 * WRITE_DATA use the same prefix for their `request_id`: all three are opaque UUID-correlated
 * binary streams and therefore share one deliberately small decoder.
 */
export const SESSION_ID_BYTES = 36;
export const MAX_DATA_BYTES = MAX_FRAME_PAYLOAD_BYTES - SESSION_ID_BYTES;

const FRAME_HEADER_BYTES = 5;
const KNOWN_TAGS = new Set<number>(Object.values(FRAME_TAGS));
const SESSION_ID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
const EMPTY = Buffer.alloc(0);

/** Any framing violation is terminal for the connection: the peer is not speaking our protocol. */
export class FramingError extends Error {}

export interface Frame {
  readonly tag: FrameTag;
  readonly payload: Buffer;
}

export function isFrameTag(value: number): value is FrameTag {
  return KNOWN_TAGS.has(value);
}

export function isSessionId(value: string): boolean {
  return SESSION_ID_PATTERN.test(value);
}

export function encodeFrame(tag: FrameTag, payload: Uint8Array = EMPTY): Buffer {
  if (payload.byteLength > MAX_FRAME_PAYLOAD_BYTES) throw new FramingError('frame payload exceeds the wire limit');
  const frame = Buffer.allocUnsafe(FRAME_HEADER_BYTES + payload.byteLength);
  frame.writeUInt8(tag, 0);
  frame.writeUInt32BE(payload.byteLength, 1);
  frame.set(payload, FRAME_HEADER_BYTES);
  return frame;
}

export function encodeJsonFrame(tag: FrameTag, value: Record<string, unknown>): Buffer {
  return encodeFrame(tag, Buffer.from(JSON.stringify(value), 'utf8'));
}

export function encodeDataFrame(tag: FrameTag, sessionId: string, data: Uint8Array): Buffer {
  if (!isSessionId(sessionId)) throw new FramingError('data frames require a lowercase dashed session UUID');
  if (data.byteLength > MAX_DATA_BYTES) throw new FramingError('data payload exceeds the wire limit');
  const payload = Buffer.allocUnsafe(SESSION_ID_BYTES + data.byteLength);
  payload.write(sessionId, 0, SESSION_ID_BYTES, 'ascii');
  payload.set(data, SESSION_ID_BYTES);
  return encodeFrame(tag, payload);
}

export interface DataFrame {
  readonly sessionId: string;
  readonly data: Buffer;
}

export function decodeDataFrame(payload: Buffer): DataFrame {
  if (payload.byteLength < SESSION_ID_BYTES) throw new FramingError('data frame is shorter than a session id');
  const sessionId = payload.subarray(0, SESSION_ID_BYTES).toString('ascii');
  if (!isSessionId(sessionId)) throw new FramingError('data frame carries an invalid session id');
  return { sessionId, data: payload.subarray(SESSION_ID_BYTES) };
}

/** JSON frames always carry an object; arrays and scalars are rejected before any field read. */
export function decodeJsonFrame(payload: Buffer): Record<string, unknown> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(payload.toString('utf8'));
  } catch {
    throw new FramingError('frame payload is not valid JSON');
  }
  if (parsed === null || typeof parsed !== 'object' || Array.isArray(parsed)) {
    throw new FramingError('frame payload is not a JSON object');
  }
  return parsed as Record<string, unknown>;
}

/**
 * Incremental decoder: TCP hands us arbitrary fragmentation, so every call may complete zero,
 * one or many frames and may leave a partial header behind.
 */
export class FrameDecoder {
  private pending: Buffer = EMPTY;

  push(chunk: Uint8Array): Frame[] {
    this.pending = this.pending.byteLength === 0 ? Buffer.from(chunk) : Buffer.concat([this.pending, chunk]);
    const frames: Frame[] = [];
    let offset = 0;
    for (;;) {
      if (this.pending.byteLength - offset < FRAME_HEADER_BYTES) break;
      const tag = this.pending.readUInt8(offset);
      const length = this.pending.readUInt32BE(offset + 1);
      if (length > MAX_FRAME_PAYLOAD_BYTES) throw new FramingError('frame length exceeds the wire limit');
      if (!isFrameTag(tag)) throw new FramingError('unknown frame tag');
      const end = offset + FRAME_HEADER_BYTES + length;
      if (this.pending.byteLength < end) break;
      frames.push({ tag, payload: Buffer.from(this.pending.subarray(offset + FRAME_HEADER_BYTES, end)) });
      offset = end;
    }
    this.pending = offset === this.pending.byteLength ? EMPTY : Buffer.from(this.pending.subarray(offset));
    return frames;
  }
}

File: ops/pty-agent/cauce_pty_agent/framing.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import re
import struct
import sys
from typing import Any

# Fail-closed exit reserved for identity violations. The container adapter supervisor already
# refuses a root runtime identity with 78; the PTY agent inherits the same contract so the
# systemd template can treat it as a permanent (non-restartable) failure.
PERMANENT_EXIT = 78
ARGUMENT_EXIT = 2

AGENT_PROTOCOL_VERSION = 1


# Wire tags. The relay implements the mirror image of this table; every value is a single byte
# and the length prefix is 4 bytes big-endian, so a frame header is always exactly 5 bytes.
TAG_AGENT_HELLO = 0x01
TAG_HELLO_ACK = 0x02
TAG_OPEN = 0x10
TAG_OPEN_OK = 0x11
TAG_OPEN_ERR = 0x12
TAG_STDIN = 0x20
TAG_STDOUT = 0x21
TAG_RESIZE = 0x22
# DA/DSR response emitted by the emulator. A dedicated tag prevents confusing it with keyboard/paste.
TAG_TERMINAL_RESPONSE = 0x23
# Backpressure per SESSION. The relay never pauses the multiplexed TLS: that would block PONG
# and the other terminals of the same agent.
TAG_PAUSE_OUTPUT = 0x24
TAG_RESUME_OUTPUT = 0x25
TAG_CLOSE = 0x30
TAG_CLOSED = 0x31
TAG_PING = 0x40
TAG_PONG = 0x41
# Reading governance files (CLAUDE.md / AGENTS.md and the memory index). A loose transaction,
# not a session: that is why it does not reuse TAG_OPEN, which opens a PTY with state.
TAG_READ = 0x50
TAG_READ_OK = 0x51
TAG_READ_ERR = 0x52
TAG_READ_DATA = 0x53
# Governed write v1. Tags kept separate from READ preserve compatibility: a relay only sends them
# when the hello advertises `write_governance_v1`; an older agent keeps terminal and read working.
TAG_WRITE = 0x54
TAG_WRITE_DATA = 0x55
TAG_WRITE_OK = 0x56
TAG_WRITE_ERR = 0x57
TAG_WRITE_CANCEL = 0x58
TAG_WRITE_BATCH = 0x59
TAG_WRITE_BATCH_DATA = 0x5A
TAG_WRITE_BATCH_OK = 0x5B
TAG_WRITE_BATCH_ERR = 0x5C
TAG_WRITE_BATCH_CANCEL = 0x5D
# Unambiguous close of a successful read. READ_ERR is already terminal on its own; READ_DONE
# only appears after READ_OK and all its READ_DATA, even when the index/directive is empty.
TAG_READ_DONE = 0x5E

MAX_FRAME = 65536
# DATA frames carry the 36 ASCII bytes of the session UUID before the raw bytes.
SESSION_ID_BYTES = 36
MAX_DATA = MAX_FRAME - SESSION_ID_BYTES
DATA_TAGS = frozenset({TAG_STDIN, TAG_STDOUT, TAG_TERMINAL_RESPONSE})
# Tags whose payload starts with a 36-byte ASCII identifier. READ_DATA carries the request's
# `request_id`, not a session, but the prefix format is intentionally the same: so the relay
# reuses its data-frame decoder without a second code path.
PREFIXED_TAGS = DATA_TAGS | frozenset({TAG_READ_DATA, TAG_WRITE_DATA, TAG_WRITE_BATCH_DATA})


CLOCK_SKEW = 5.0


SESSION_ID_RE = re.compile(r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$")
TICKET_RE = re.compile(r"^v1\.([A-Za-z0-9_-]{1,4096})\.([A-Za-z0-9_-]{43})$")
IDENTITY_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$")
MODES = ("shell", "harness")


class PermanentError(RuntimeError):
    """The bundle or the runtime identity is wrong; retrying cannot help."""


class ProtocolError(RuntimeError):
    """The peer sent something the framing layer refuses; the connection is dropped."""


class TicketError(RuntimeError):
    """The OPEN ticket did not verify. `reason` travels back in OPEN_ERR, nothing else does."""

    def __init__(self, reason: str, detail: str = "") -> None:
        super().__init__(reason if not detail else f"{reason}:{detail}")
        self.reason = reason
        self.detail = detail


def fail(message: str, code: int = ARGUMENT_EXIT) -> None:
    print(message, file=sys.stderr)
    raise SystemExit(code)


def log(message: str) -> None:
    # stderr only, and never a secret: the launcher's journal is readable by anyone who can read
    # the unit, so only names, paths, sizes and identities are printed.
    print(f"cauce-pty-agent: {message}", file=sys.stderr, flush=True)


# ---------------------------------------------------------------------------------------------
# Framing
# ---------------------------------------------------------------------------------------------


def encode_frame(tag: int, payload: bytes) -> bytes:
    if not 0 <= tag <= 0xFF:
        raise ProtocolError("frame tag is out of range")
    if len(payload) > MAX_FRAME:
        raise ProtocolError("frame payload exceeds the negotiated maximum")
    return struct.pack("!BI", tag, len(payload)) + payload


def encode_json(tag: int, document: dict[str, Any]) -> bytes:
    return encode_frame(tag, json.dumps(document, separators=(",", ":"), sort_keys=True).encode("utf-8"))


def encode_data(tag: int, session_id: str, data: bytes) -> bytes:
    if tag not in PREFIXED_TAGS:
        raise ProtocolError("only data/response frames carry a 36 byte prefix")
    identifier = session_id.encode("ascii")
    if len(identifier) != SESSION_ID_BYTES:
        raise ProtocolError("session id must be a 36 byte UUID")
    if len(data) > MAX_DATA:
        raise ProtocolError("data chunk exceeds the per-frame maximum")
    return encode_frame(tag, identifier + data)


def decode_data(payload: bytes) -> tuple[str, bytes]:
    if len(payload) < SESSION_ID_BYTES:
        raise ProtocolError("data frame is shorter than its session prefix")
    identifier = payload[:SESSION_ID_BYTES].decode("ascii", "replace")
    if not SESSION_ID_RE.fullmatch(identifier):
        raise ProtocolError("data frame carries an invalid session id")
    return identifier, payload[SESSION_ID_BYTES:]


class FrameDecoder:
    """Incremental decoder. Survives arbitrary fragmentation, including one byte at a time."""

    def __init__(self) -> None:
        self._buffer = bytearray()

    def feed(self, chunk: bytes) -> list[tuple[int, bytes]]:
        self._buffer.extend(chunk)
        frames: list[tuple[int, bytes]] = []
        while True:
            if len(self._buffer) < 5:
                return frames
            tag = self._buffer[0]
            length = struct.unpack_from("!I", self._buffer, 1)[0]
            if length > MAX_FRAME:
                raise ProtocolError("peer announced a frame above the negotiated maximum")
            if len(self._buffer) < 5 + length:
                return frames
            frames.append((tag, bytes(self._buffer[5:5 + length])))
            del self._buffer[:5 + length]


# ---------------------------------------------------------------------------------------------
# Ticket verification
# ---------------------------------------------------------------------------------------------


def b64url_decode(text: str) -> bytes:
    # Unpadded base64url exactly as the gateway emits it: '-' and '_' are the URL-safe aliases of
    # '+' and '/', and the padding is restored here because base64 refuses a short final quantum.
    normalized = text.replace("-", "+").replace("_", "/")
    decoded = base64.b64decode(normalized + "=" * (-len(normalized) % 4))
    canonical = base64.urlsafe_b64encode(decoded).decode("ascii").rstrip("=")
    if canonical != text:
        raise ValueError("base64url segment is not canonical")
    return decoded


def verify_ticket(alias_key: bytes, ticket: str, now: float) -> dict[str, Any]:
    """Second, independent check: a compromised relay still cannot open a shell here.

    The relay already verified this ticket; the agent verifies it again with the per-alias key that
    only exists inside this container, so the relay is not trusted to name a target.
    """
    if not isinstance(ticket, str):
        raise TicketError("ticket_malformed")
    match = TICKET_RE.fullmatch(ticket)
    if not match:
        raise TicketError("ticket_malformed")
    encoded_payload, encoded_mac = match.group(1), match.group(2)
    try:
        signature = b64url_decode(encoded_mac)
    except ValueError:  # binascii.Error is a ValueError subclass
        # Do not expose whether a signature had bad alphabet, length or unused trailing bits.
        raise TicketError("ticket_bad_signature") from None
    expected = hmac.new(alias_key, ("v1." + encoded_payload).encode("ascii"), hashlib.sha256).digest()
    if len(signature) != len(expected) or not hmac.compare_digest(expected, signature):
        raise TicketError("ticket_bad_signature")
    try:
        payload = json.loads(b64url_decode(encoded_payload).decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        raise TicketError("ticket_malformed") from None
    if not isinstance(payload, dict) or payload.get("v") != 1:
        raise TicketError("ticket_malformed")
    expires = payload.get("exp")
    if not isinstance(expires, (int, float)) or isinstance(expires, bool):
        raise TicketError("ticket_malformed")
    if now > float(expires) + CLOCK_SKEW:
        raise TicketError("ticket_expired")
    return payload


def authorize_ticket(payload: dict[str, Any], identity: dict[str, Any], session_id: str) -> str:
    """Every target field must name THIS container generation; anything else is refused."""
    if payload.get("sid") != session_id:
        raise TicketError("session_mismatch")
    target = payload.get("tgt")
    if not isinstance(target, dict):
        raise TicketError("ticket_malformed")
    for field, expected in (
        ("tenant", identity["tenant_id"]),
        ("alias", identity["alias"]),
        ("container", identity["container_id"]),
        ("generation", identity["generation"]),
        ("uid", identity["runtime_uid"]),
    ):
        if target.get(field) != expected:
            raise TicketError("target_mismatch", f"tgt.{field}")
    mode = payload.get("mode")
    if mode not in MODES:
        raise TicketError("mode_unknown")
    return mode
